### ISSUE TITLE
Revisit how Services are deployed - do not rely on Template

### ISSUE THREAD

**Issue by vpavlin** [opendatahub-io/jupyterhub-singleuser-profiles#19]
Currently a service is deployed by processing a template (https://github.com/vpavlin/jupyterhub-singleuser-profiles/blob/master/jupyterhub_singleuser_profiles/service.py#L57). Since templates are going away we need to figure out a different way.

We could probably include the service resource directly in the singleuser configMap - for Spark that would mean:

Current:

```
          spark:
            template: 'jupyterhub-spark-operator-configmap'
            parameters:
              WORKER_NODES: '2'
              MASTER_NODES: '1'
              MASTER_MEMORY_LIMIT: '2Gi'
              MASTER_CPU_LIMIT: '1'
              MASTER_MEMORY_REQUEST: '2Gi'
              MASTER_CPU_REQUEST: '1'
              WORKER_MEMORY_LIMIT: '2Gi'
              WORKER_CPU_LIMIT: '1'
              WORKER_MEMORY_REQUEST: '2Gi'
              WORKER_CPU_REQUEST: '1'
              SPARK_IMAGE: 'quay.io/opendatahub/spark-cluster-image:spark22python36'
            return:
              SPARK_CLUSTER: 'metadata.name'
```

Future:

```
spark:
  resources:
  - apiVersion: v1
    kind: ConfigMap
    metadata:
      labels:
        radanalytics.io/kind: SparkCluster
      name: spark-cluster-${USERNAME}
    data:
      config: |-
        worker:
          instances: 2
          memoryLimit: 2Gi
          cpuLimit: 1
          memoryRequest: 2Gi
          cpuRequest: 1
        master:
          instances: 1
          memoryLimit: 2Gi
          cpuLimit: 1
          memoryRequest: 2Gi
          cpuRequest: 1
        customImage: 'quay.io/opendatahub/spark-cluster-image:spark22python36'
        env:
        - name: SPARK_METRICS_ON
          value: prometheus
  return:
    SPARK_CLUSTER: 'metadata.name'
```

We could also link a configMap/secret containing the resource

Thoughts, @lavlas @erikerlandson

**Comment by LaVLaS** [comment]
I like the idea of moving the template to a configmap outiside of the singleuserprofile (maybe a singleuserservice?).  This would allow it to be re-used across multiple profiles without having duplicate it for any profile that wants to change the some values.  The big downside is jupyterhub would need to support some type of template (jinja2?) process for variables.

**Comment by vpavlin** [comment]
We can use jinja or jsonpath. I think that using jsonpath would give us more flexibility. 

```
spark:
  resource:
    kind: ConfigMap
    name: spark-cluster-template
  configuration:
    data.worker.instances: 1
    data.worker.memoryLimit: 4Gi
    data.master.memoryLimit: 1Gi
  return:
    SPARK_CLUSTER: 'metadata.name'
```

**Comment by vpavlin** [comment]
or if we decide for Jinja2

for

```
  - apiVersion: v1
    kind: ConfigMap
    metadata:
      labels:
        radanalytics.io/kind: SparkCluster
      name: spark-cluster-${USERNAME}
    data:
      config: |-
        worker:
          instances: ${worker_instances}
          memoryLimit: ${worker_memory_limit}
          cpuLimit: 1
          memoryRequest: 2Gi
          cpuRequest: 1
        master:
          instances: 1
          memoryLimit: ${master_memory_limit}
          cpuLimit: 1
          memoryRequest: 2Gi
          cpuRequest: 1
        customImage: 'quay.io/opendatahub/spark-cluster-image:spark22python36'
        env:
        - name: SPARK_METRICS_ON
          value: prometheus
```


```
spark:
  resources:
  - kind: ConfigMap
    name: spark-cluster-template
  configuration:
    worker_instances: 1
    worker_memory_limit: 4Gi
    master_memory_limit: 1Gi
  return:
    SPARK_CLUSTER: 'metadata.name'
```

**Comment by vpavlin** [comment]
Since we are moving to CustomResources for Spark Operator we have evolved this a bit more:

```
kind: ConfigMap
metadata:
  name: jupyter-services-template
...
data:
  sparkClusterTemplate: |
      kind: SparkCluster
      ....
  airflowCluster: |
      kind: AirflowCluster
      ...
  airflowBase: |
      kind: AirflowBase
      ....
```

And the singleuser-profile CM would reference:

```
spark:
  resources:
  - name: jupyter-services-template
    path: sparkClusterTemplate
```

and 

```
airflow:
  resources:
  - name: jupyter-services-template
    path: airflowCluster
  - name: jupyter-services-template
    path: airflowBase
```
```

**Comment by vpavlin** [comment]
Fixed by #26
